include/envoy/registry/registry.h

```diff
@@ -118,15 +118,22 @@ class FactoryCategoryRegistry : public BaseFactoryCategoryRegistry {
   }
 
 private:
-  // Allow RegisterFactory to register a category, but no-one else.
-  // This enforces correct use of the registration machinery.
+  // Allow RegisterFactory and the test helper InjectFactoryCategory to register a category, but
+  // no-one else. This enforces correct use of the registration machinery.
   template <class T, class Base> friend class RegisterFactory;
+  template <class Base> friend class InjectFactoryCategory;
 
-  static void registerCategory(const std::string& category, FactoryRegistryProxy* factoryNames) {
-    auto result = factories().emplace(std::make_pair(category, factoryNames));
+  static void registerCategory(const std::string& category, FactoryRegistryProxy* factory_names) {
+    auto result = factories().emplace(std::make_pair(category, factory_names));
     RELEASE_ASSERT(result.second == true,
                    fmt::format("Double registration for category: '{}'", category));
   }
+
+  static void deregisterCategoryForTest(const std::string& category) {
+    factories().erase(category);
```

Comment:
I think what you've built overlaps and is harmonious with Envoy::Test::Globals (https://github.com/envoyproxy/envoy/blob/master/test/test_common/global.h). What that could add, if you integrated it, is a between-test-method check that you'd cleaned everything up, and the ability to register a factory in the test-code where it's needed, even if was already registered at the test's top level.

Up to you, and I don't think that should block this PR; just calling that infrastructure to your attention.